HPCC-27987 Add ability to configure default dfuserver options globally (via environment), e.g. nocommon/maxConnections

AttilaVamos opened the pull request:
Implement
- FileServices.GetNoCommonDefault();
function to read default value from environment.xml in bare-metal system.

Add default values to default environment.xml.in .

Replace noCommon default value with GetNoCommonDefault() function call in:
- SprayFixed
- SprayVariable
- SprayDelimited
- SprayXml
- SprayJson

Add debug log into DFU Server spray to check the value of 'noCommon' and 'maxConnections' parameters.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3128,3 +3128,49 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+
+bool getDefaultValue(const char * processName, const char * propertyName, StringBuffer & strDefaultValue)
+{
+    Owned<IConstEnvironment> daliEnv = openDaliEnvironment();
```
This shouldn't be called in (is undefined) in containerized version.

AttilaVamos replied:
Yes, that is true. I move it into a conditional block.

ghalliday commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3128,3 +3128,49 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+
+    if (env.get())
+    {
+        Owned<IPropertyTreeIterator> processIter = env->getElements(processName);
```
If these are global defaults, does it make sense to use the values associated with the 1st dfu server?

What is the corresponding implementation for containerized mode?  

(I'm not sure I have a good answer to either question.  @jakesmith ideas?)

AttilaVamos replied:
My original thought/implementation was that, but then it raised several questions/implications, therefore I moved them from Instance to Server Process.

AttilaVamos replied:
Changed

ghalliday commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3128,3 +3128,49 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+    return true;
+}
+
+FILESERVICES_API int FILESERVICES_CALL fsGetMaxConnectionsDefault()
```
Why is this needed, rather than dfuserver using a global default value?  It is already only set in the dfu request if it has been specified.

AttilaVamos replied:
@jakesmith ?

jakesmith replied:
Agree, for this and any other property that can, it should pickup the default within dfuserver when it's processing the dfu workunit.
NB: it already defaults there to a hard-wired DEFAULT_MAX_CONNECTIONS, so that can be switched for the global default if configured.

Unfortunately, the same cannot be applied to boolean flags (e.g. noCommon) where it's default can not be a 'unset' value and therefore cannot be tested in dfuserver in order to use a global default.
For properties that can't (e.g. noCommon), the current approach of picking up a configured global default and passing it in, is the best approach I can think of.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+        if (env.get())
+        {
+            Owned<IPropertyTreeIterator> processIter = env->getElements(processName);
+            if (processIter->first())
```
> Gavin Halliday:
> If these are global defaults, does it make sense to use the values associated with the 1st dfu server?
> 

> Attila Vamos:
> My original thought/implementation was that, but then it raised several questions/implications, therefore I moved them from Instance to Server Process.

They should not live in the 1st DfuServerProcess, there may be multiple (not just multiple Instance's under each DfuServerProcess).
The original idea was to just allow these values to be set globally, not per dfuserver.
These settings should move into a global section of the environment (maybe EnvSettings, or could add new Software/Globals section).

AttilaVamos replied:
Done

jakesmith replied:
minor: The above code implies there could be multiple Globals sections, which doesn't make a great deal of sense.
Better to change not to iterate..

jakesmith replied:
@AttilaVamos - Not sure if you've missed my previous comments, re. please leave the "Resolve conversation" button for use by the reviewers only:
https://github.com/hpcc-systems/HPCC-Platform/pull/16528#discussion_r975491777
https://github.com/hpcc-systems/HPCC-Platform/pull/16528#discussion_r977883966

jakesmith commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -217,7 +217,9 @@
-                    queue="dfuserver_queue">
+                    queue="dfuserver_queue"
+                    noCommon="true"
+                    maxConnections="-1">
```
See other comment, should be in a global section of the environment.xml..

AttilaVamos replied:
Moved

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+    // Othervise it returns with true, the recent default value of 'noCommon'
+
+    StringBuffer strDefaultValue;
+    if (getDefaultValue("Software/DfuServerProcess", "@noCommon", strDefaultValue))
```
this could be called quite often and it's not that cheap (openDaliEnvironment() etc.).
To avoid the cost per call, you could use a singleton pattern to ensure the action is only done on 1st call only.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+{
+    // Returns with true if a default value retrieved and false if not.
+
+    #ifdef _CONTAINERIZED
```
style: we don't normally indent #if's/#endif's

AttilaVamos replied:
Changed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,56 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+    // Returns with true if a default value retrieved and false if not.
+
+    #ifdef _CONTAINERIZED
+        // TBD
```
I think the implementation of this should be trivial.
e.g. in fsGetNoCommonDefault():
```
    return getComponentConfigSP()->getPropBool("@noCommon", true);
```

but should be tested.

AttilaVamos replied:
Changed

jakesmith reviewed (changes requested):
@AttilaVamos - please see comments

AttilaVamos commented on dali/ft/filecopy.cpp:
```diff
@@ -3087,6 +3087,8 @@ void FileSprayer::spray()
     LOG(MCdebugInfo, job, "compressedInput:%d, compressOutput:%d", compressedInput, compressOutput);
     LOG(MCdebugInfo, job, "noCommon:%s", boolToStr(options->getPropBool(ANnocommon)));
+    LOG(MCdebugInfo, job, "maxConnections:%d", options->getPropInt(ANmaxConnections));
+
```
Removed

jakesmith replied:
@AttilaVamos - not important in this trivial case of course, but as per code review doc. :https://github.com/hpcc-systems/HPCC-Platform/blob/e33ff93f7bbd4938fafc085f0cadadbec0396b04/devdoc/CodeReviews.md?plain=1#L46
please leave the 'Resolve conversation' button(s) for the reviewers, it helps the reviewer track issues

AttilaVamos commented on ecllibrary/std/File.ecl:
```diff
@@ -105,6 +105,30 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
+ * @return              Integer value of 'maxConnection' property.
+ */
+
+EXPORT INTEGER4 GetMaxCOnnectionsDefault() := 
```
Changed.

AttilaVamos commented on ecllibrary/std/File.ecl:
```diff
@@ -105,6 +105,30 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
+    lib_fileservices.FileServices.GetNoCommonDefault();
+
+/**
+ * Returns the integer value of maxConnetcions property of the first DFUServer instance if it is defined,
```
Changed

AttilaVamos reviewed (commented):
Changed

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -105,6 +105,16 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
 /*------------------------------------- Spray functions -----------------------------------------------------------*/
 
+/**
+ * Returns the boolean value of 'noCommon' property of the first DFUServer instance if it is defined,
```
minor: comment is out of date (settings now in Global)

AttilaVamos replied:
Changed

ghalliday replied:
@AttilaVamos you marked this as resolved (you shouldn't be doing that as Jake has mentioned), and you haven't changed the comment.

jakesmith commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -20,6 +20,7 @@
         storageWrites="0.0038"/>
  </Hardware>
  <Software>
+  <Globals  noCommon="true" />
```
trivial: extra space before />

Also, where has maxConnections definition gone?
The comments/discussion said the default should be picked up by dfuserver. Its default value (and any other non-binary) can be better and more easily picked up at the server, and the environment default if configured set then.

AttilaVamos replied:
Removed

ghalliday replied:
Again you have marked as resolved, but not released the fix.

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+#ifdef _CONTAINERIZED
+    return getComponentConfigSP()->getPropBool("@noCommon", true);
+#else
+    CriticalSection  noCommonDefCrit;
```
This CS can't be a local stack member. It will not protect any other re-entrant call, since they'll all have their own CS they can lock, needs to be a global that any thread coming here locks to lock others out.

AttilaVamos replied:
Changed

jakesmith reviewed (changes requested):
@AttilaVamos - please see comments.

Also maxConnections functionality seems to have been removed completely?
but the previous comments and discussions were re. picking up the configured default in dfuserver..

jakesmith commented on ecllibrary/std/File.ecl:
```diff
@@ -105,6 +105,16 @@ EXPORT FsLandingZoneRecord := lib_fileservices.FsLandingZoneRecord;
 
+/**
+ * Returns the boolean value of 'noCommon' property of the first DFUServer instance if it is defined,
+ * othervise return with 'true' as default.
```
typo: otherwise

AttilaVamos replied:
Changed

ghalliday reviewed (changes requested):
@AttilaVamos doesn't look like you have addressed the comments from Jake.  Please re-request review once you think it is ready.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -2491,6 +2491,37 @@ bool FileSprayer::needToCalcOutput()
 }
 
+
+CriticalSection  maxConnectionsDefCrit;
```
minor: should be a static.

trivial: double space before max.

AttilaVamos replied:
Changed

jakesmith commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -20,6 +20,8 @@
         storageWrites="0.0038"/>
  </Hardware>
  <Software>
+  <Globals  noCommon="true"
```
trivial: 2 spaces

AttilaVamos replied:
Changed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
         throw makeStringExceptionV(0, "ClearExpireDays: Could not find logical file %s", lfn.str());
 }
+
+CriticalSection  noCommonDefCrit;
```
minor: should be static

trivial: 2 spaces.

AttilaVamos replied:
Changed

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -2491,6 +2491,37 @@ bool FileSprayer::needToCalcOutput()
+
+CriticalSection  maxConnectionsDefCrit;
+static unsigned maxConnectionsDef = NotFound;
+unsigned getMaxConnectionsDefault()
```
minor: this could be static

AttilaVamos replied:
Changed

jakesmith reviewed (changes requested):
@AttilaVamos - as discussed offline, I think this default (and others if we ever add them), belong in dfuserver (dfurun), not filecopy.
options are passed in from dfuserver through the sprayer it uses, and if set there (if not already set), all places that reference ANmaxConnections will pick up the env. default

jakesmith commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -20,6 +20,8 @@
  </Hardware>
  <Software>
+  <Globals noCommon="true"
+           maxConnections="0"/>
```
should be -1 as per see file.ecl.

AttilaVamos replied:
Changed

AttilaVamos committed:
HPCC-27987 Add ability to configure default dfuserver options globally (via environment), e.g. nocommon/maxConnections

Implement
- FileServices.GetNoCommonDefault();
function to read default value from environment.xml in bare-metal system.

Add default values to default environment.xml.in .

Replace noCommon default value with GetNoCommonDefault() function call in:
- SprayFixed
- SprayVariable
- SprayDelimited
- SprayXml
- SprayJson

Add debug log into DFU Server spray to check the value of 'noCommon' and
'maxConnections' parameters.

Tested manually.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
         throw makeStringExceptionV(0, "ClearExpireDays: Could not find logical file %s", lfn.str());
 }
+
+static CriticalSection  noCommonDefCrit;
```
trivial: still 2 spaces.

AttilaVamos replied:
Changed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,30 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+        if (env.get())
+        {
+            Owned<IPropertyTree> globals = env->getPropTree("Software/Globals");
+            noCommonDef = globals->getPropBool("@noCommon", noCommonDef);
```
trivial: as per code in dfurun, it is unecessary to fetch tree into a linked object, and then the fetch the bool from it. It would be simpler/cleaner to be:
```
    noCommonDef = env->getPropBool("@noCommon", noCommonDef);
```

AttilaVamos replied:
Changed

jakesmith reviewed (approved):
looks ready.

jakesmith commented:
@ghalliday - please sanity check

AttilaVamos commented:
@ghalliday Should I do anything more for this PR? The users are very keen to use it.

ghalliday commented on dali/dfu/dfurun.cpp:
```diff
@@ -1117,6 +1117,26 @@ class CDFUengine: public CInterface, implements IDFUengine
+        if (!opttree->hasProp("@maxConnections"))
+        {
+            int configDefault = -1;
+            if (isContainerized())
```
This is the one situation where you need to use #ifdef _CONTAINERIZED rather than isContainerized().  

The containerized version will not define getEnvironmentFactory(), so you will get link errors/warnings if symbol resolution is enforced at link time.

AttilaVamos replied:
I see. Changed.

ghalliday commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -3140,3 +3140,27 @@ FILESERVICES_API void FILESERVICES_CALL fsClearExpireDays(ICodeContext * ctx, co
+            noCommonDef = env->getPropBool("@noCommon", noCommonDef);
+    }
+#endif
+    return noCommonDef;
```
minor: Should explicitly cast to boolean, otherwise you'll get warnings from compilers (especially windows)

AttilaVamos replied:
Changed

ghalliday commented on plugins/proxies/lib_fileservices.ecllib:
```diff
@@ -112,4 +112,5 @@ export FileServices := SERVICE : plugin('fileservices'), time
   integer4 GetExpireDays(const varstring lfn) : c,context,entrypoint='fsGetExpireDays';
   SetExpireDays(const varstring lfn, integer4 expireDays) : c,context,entrypoint='fsSetExpireDays';
   ClearExpireDays(const varstring lfn) : c,context,entrypoint='fsClearExpireDays';
+  boolean GetNoCommonDefault() : c,context,entrypoint='fsGetNoCommonDefault';
```
I'm not quite sure how this links, probably because it is within an extern "C" so name mangling is disabled.  The function does not have an ICodeContext  * parameter, so it should not have a context attribute in this list.

AttilaVamos replied:
Removed

ghalliday commented on dali/ft/filecopy.cpp:
```diff
@@ -2500,14 +2500,20 @@ unsigned FileSprayer::numParallelConnections(unsigned limit)
+    LOG(MCdebugInfo, job, "In numParallelSlaves():");
+    LOG(MCdebugInfo, job, "    numPullers:%u", numPullers);
+    LOG(MCdebugInfo, job, "    maxConnections:%u", maxConnections);
+    LOG(MCdebugInfo, job, "    connectOption:%u", connectOption);
```
Slightly concerned about the extra logging. does it come out in the dfuserver log by default?  If so it is too verbose.

AttilaVamos replied:
I see. I put them in #ifdef _DEBUG .. #endif block.

ghalliday reviewed (changes requested):
@AttilaVamos a few problems remaining, and a question.

ghalliday commented:
@AttilaVamos please don't squash changes until the review has accepted them.

ghalliday reviewed (approved):
I think it is now ook to merge.
Slightly uneasy about
(1) using Software/Globals rather than copying the global section into the configuration for this component (as is done with costs) and also not cached.
(2) still additional logging.  I think this is can be reviewed as part of a more comprehensive review of logging.

ghalliday commented:
@AttilaVamos Github has a helpful link to compare the two versions which helped.

AttilaVamos commented:
@ghalliday Thanks and sorry again.

AttilaVamos commented:
@richardkchapman please merge.

richardkchapman commented:
Merged to master and backmerged to 8.10.x for 8.10.0
